### LOAD DATASET 

In [31]:
import pandas as pd
import numpy as np
SuperStore= pd.read_csv("C:/Users/ayoub/Desktop/janvier-2027/python w2/projet/data/superstore.csv" ,sep=',' )
SuperStore

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales
0,1,CA-2017-152156,08/11/2017,11/11/2017,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600
1,2,CA-2017-152156,08/11/2017,11/11/2017,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400
2,3,CA-2017-138688,12/06/2017,16/06/2017,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036.0,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200
3,4,US-2016-108966,11/10/2016,18/10/2016,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775
4,5,US-2016-108966,11/10/2016,18/10/2016,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9795,9796,CA-2017-125920,21/05/2017,28/05/2017,Standard Class,SH-19975,Sally Hughsby,Corporate,United States,Chicago,Illinois,60610.0,Central,OFF-BI-10003429,Office Supplies,Binders,"Cardinal HOLDit! Binder Insert Strips,Extra St...",3.7980
9796,9797,CA-2016-128608,12/01/2016,17/01/2016,Standard Class,CS-12490,Cindy Schnelling,Corporate,United States,Toledo,Ohio,43615.0,East,OFF-AR-10001374,Office Supplies,Art,"BIC Brite Liner Highlighters, Chisel Tip",10.3680
9797,9798,CA-2016-128608,12/01/2016,17/01/2016,Standard Class,CS-12490,Cindy Schnelling,Corporate,United States,Toledo,Ohio,43615.0,East,TEC-PH-10004977,Technology,Phones,GE 30524EE4,235.1880
9798,9799,CA-2016-128608,12/01/2016,17/01/2016,Standard Class,CS-12490,Cindy Schnelling,Corporate,United States,Toledo,Ohio,43615.0,East,TEC-PH-10000912,Technology,Phones,Anker 24W Portable Micro USB Car Charger,26.3760


### PRINTING DATASET INFORMATIONS 


In [32]:
print("the number of rows is :" , SuperStore.shape[0] )
print("the number of columns is :" , SuperStore.shape[1] )
print( SuperStore.info(memory_usage=False))

the number of rows is : 9800
the number of columns is : 18
<class 'pandas.DataFrame'>
RangeIndex: 9800 entries, 0 to 9799
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9800 non-null   int64  
 1   Order ID       9800 non-null   str    
 2   Order Date     9800 non-null   str    
 3   Ship Date      9800 non-null   str    
 4   Ship Mode      9800 non-null   str    
 5   Customer ID    9800 non-null   str    
 6   Customer Name  9800 non-null   str    
 7   Segment        9800 non-null   str    
 8   Country        9800 non-null   str    
 9   City           9800 non-null   str    
 10  State          9800 non-null   str    
 11  Postal Code    9789 non-null   float64
 12  Region         9800 non-null   str    
 13  Product ID     9800 non-null   str    
 14  Category       9800 non-null   str    
 15  Sub-Category   9800 non-null   str    
 16  Product Name   9800 non-null   str    
 17  Sale

### First and last  5 elements 


In [33]:
print(SuperStore.head())
print("------")
print(SuperStore.tail())

   Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
0       1  CA-2017-152156  08/11/2017  11/11/2017    Second Class    CG-12520   
1       2  CA-2017-152156  08/11/2017  11/11/2017    Second Class    CG-12520   
2       3  CA-2017-138688  12/06/2017  16/06/2017    Second Class    DV-13045   
3       4  US-2016-108966  11/10/2016  18/10/2016  Standard Class    SO-20335   
4       5  US-2016-108966  11/10/2016  18/10/2016  Standard Class    SO-20335   

     Customer Name    Segment        Country             City       State  \
0      Claire Gute   Consumer  United States        Henderson    Kentucky   
1      Claire Gute   Consumer  United States        Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States      Los Angeles  California   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   

   Postal Code Region       Product ID         Cat

#### Columns to be changed
We can change ship date from str to a datetime object  rather than a string
and we can change the postal code into an integer rather than float

### STATS

In [34]:
print(SuperStore.describe())

            Row ID   Postal Code         Sales
count  9800.000000   9789.000000   9800.000000
mean   4900.500000  55273.322403    230.769059
std    2829.160653  32041.223413    626.651875
min       1.000000   1040.000000      0.444000
25%    2450.750000  23223.000000     17.248000
50%    4900.500000  58103.000000     54.490000
75%    7350.250000  90008.000000    210.605000
max    9800.000000  99301.000000  22638.480000


### Missing value

In [35]:
print(SuperStore.isnull().sum())
print("the percentage of missing value in postal code is :" , ((SuperStore["Postal Code"].isna().sum()) / len(SuperStore)) * 100 ,"%")
Missing=SuperStore[SuperStore["Postal Code"].isna()][["City","State" , "Postal Code"]]
print(Missing)

Row ID            0
Order ID          0
Order Date        0
Ship Date         0
Ship Mode         0
Customer ID       0
Customer Name     0
Segment           0
Country           0
City              0
State             0
Postal Code      11
Region            0
Product ID        0
Category          0
Sub-Category      0
Product Name      0
Sales             0
dtype: int64
the percentage of missing value in postal code is : 0.11224489795918367 %
            City    State  Postal Code
2234  Burlington  Vermont          NaN
5274  Burlington  Vermont          NaN
8798  Burlington  Vermont          NaN
9146  Burlington  Vermont          NaN
9147  Burlington  Vermont          NaN
9148  Burlington  Vermont          NaN
9386  Burlington  Vermont          NaN
9387  Burlington  Vermont          NaN
9388  Burlington  Vermont          NaN
9389  Burlington  Vermont          NaN
9741  Burlington  Vermont          NaN


As we can see from the data all postal code that are missing are for vermont state so it is suitable to fill the nan values with Burligtone vermont postal code which is 05401

In [36]:
SuperStore["Postal Code"] = SuperStore["Postal Code"].fillna(5401.0)
print(SuperStore.isnull().sum())
SuperStore["Postal Code"]=SuperStore["Postal Code"].astype(int).astype(str).str.zfill(5)
print("the new type of Postal code is " , SuperStore["Postal Code"].dtype)


Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
dtype: int64
the new type of Postal code is  str


### Duplicated values

In [37]:
ColumnsToCheck=SuperStore.columns.drop('Row ID')
SuperStore[SuperStore.duplicated(subset=ColumnsToCheck, keep=False)]

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales
3405,3406,US-2015-150119,23/04/2015,27/04/2015,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,Ohio,43229,East,FUR-CH-10002965,Furniture,Chairs,Global Leather Highback Executive Chair with P...,281.372
3406,3407,US-2015-150119,23/04/2015,27/04/2015,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,Ohio,43229,East,FUR-CH-10002965,Furniture,Chairs,Global Leather Highback Executive Chair with P...,281.372


because of teh Row ID value we cant have a duplicated row because it increment from 1 to 9800 however if we check the dataset without the Row ID we indeed find a repeated row 
but it is really difficult to distinguish if it is a repeated order or it is an actuall duplicate 

for our case we will consider it as repeated becaus eit is very  very unlikely that a person commit two same orders in the same day and in an addition they are consecutive in the dataset Row id = 3406 , 3407 

In [38]:
DropIndex=SuperStore[SuperStore["Row ID"] == 3407].index
SuperStore.drop(DropIndex , inplace=True)
SuperStore[SuperStore.duplicated(subset=ColumnsToCheck, keep=False)]


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales


### Data consistency

#### categorical values inconsistency 
Ship mode : we have 4 unique values  ['Second Class', 'Standard Class', 'First Class', 'Same Day'] all of these values are consistente and have meaning  

Segment :  we have 3 unique values ['Consumer', 'Corporate', 'Home Office'] this is consistent

Country : we have 1 unique values ['United States'] this is valid

State : to be considered to be consistente and because we have one country covered by this data which is america we should expect no more than 50 state which is true 

Region : ['South', 'West', 'Central', 'East']are  valid

category : ['Furniture', 'Office Supplies', 'Technology']  3 product category which are valid

Sub Category : we crossed checked category with sub category to find if there is somme inconsstencies such ass categori: fourniter-sub category = Arts which we didnot find 

thats make us conclude that our data set categorical values are consistente 

In [39]:
print(SuperStore["Ship Mode"].unique())
print(SuperStore["Segment"].unique())
print(SuperStore["Country"].unique())
print(SuperStore["State"].nunique())
print(SuperStore["Region"].unique())
print(SuperStore[SuperStore["Category"] == "Furniture"]["Sub-Category"].unique())
print(SuperStore[SuperStore["Category"] == "Office Supplies"]["Sub-Category"].unique())
print(SuperStore[SuperStore["Category"] == "Technology"]["Sub-Category"].unique())


<StringArray>
['Second Class', 'Standard Class', 'First Class', 'Same Day']
Length: 4, dtype: str
<StringArray>
['Consumer', 'Corporate', 'Home Office']
Length: 3, dtype: str
<StringArray>
['United States']
Length: 1, dtype: str
49
<StringArray>
['South', 'West', 'Central', 'East']
Length: 4, dtype: str
<StringArray>
['Bookcases', 'Chairs', 'Tables', 'Furnishings']
Length: 4, dtype: str
<StringArray>
[    'Labels',    'Storage',        'Art',    'Binders', 'Appliances',
      'Paper',  'Envelopes',  'Fasteners',   'Supplies']
Length: 9, dtype: str
<StringArray>
['Phones', 'Accessories', 'Machines', 'Copiers']
Length: 4, dtype: str


#### Cross data consistency
Order ID : for the same order we can not find different Order IDs which is true in our Data Set

Customer ID Customer Name : we must have the same number of Customer ID and Customer Name wxhich is true 

Product ID and Product Name :  we must have the same number of Product ID and   Product Name which is False  we have more ditinct Product ID than Distinc Product Name which means there are some product which have different IDs


In [40]:
ColumnsToCheck2=SuperStore.columns.drop(['Row ID' , 'Order ID'])
print("The number of duplicated orders is " , SuperStore.duplicated(subset=ColumnsToCheck2, keep=False).sum())
print(SuperStore["Customer ID"].nunique() == SuperStore["Customer Name"].nunique() )
print(SuperStore["Product ID"].nunique(),SuperStore["Product Name"].nunique() )

CountN=SuperStore.groupby("Product ID")["Product Name"].nunique().reset_index()
IDsNames=CountN[CountN["Product Name"]>1]
print(IDsNames)



The number of duplicated orders is  0
True
1861 1849
           Product ID  Product Name
18    FUR-BO-10002213             2
65    FUR-CH-10001146             2
183   FUR-FU-10001473             2
284   FUR-FU-10004017             2
290   FUR-FU-10004091             2
295   FUR-FU-10004270             2
310   FUR-FU-10004848             2
311   FUR-FU-10004864             2
387   OFF-AP-10000576             2
506   OFF-AR-10001149             2
721   OFF-BI-10002026             2
831   OFF-BI-10004632             2
832   OFF-BI-10004654             2
1044  OFF-PA-10000357             2
1049  OFF-PA-10000477             2
1064  OFF-PA-10000659             2
1086  OFF-PA-10001166             2
1144  OFF-PA-10001970             2
1156  OFF-PA-10002195             2
1168  OFF-PA-10002377             2
1198  OFF-PA-10003022             2
1331  OFF-ST-10001228             2
1417  OFF-ST-10004950             2
1516  TEC-AC-10002049             2
1532  TEC-AC-10002550             2
1568  TEC-A

as we can see here we have 60 row which have a probleme in their product Name we will standrize them for every product id x we will assigne for it the one product name y and we cant tolerate that our solution is based on the issue because if a product id has muiltiple names it is a really big probleme

In [42]:
ProblemIDs= IDsNames["Product ID"]

TargetMap = SuperStore[SuperStore['Product ID'].isin(ProblemIDs)].groupby("Product ID")["Product Name"].first().to_dict()
SuperStore.loc[SuperStore["Product ID"].isin(ProblemIDs) , "Product Name"]=SuperStore["Product ID"].map(TargetMap)

SuperStore["Product ID"].nunique()-SuperStore["Product Name"].nunique()

44

as we can see the result is -32 that means that we have a probleme in the other direction one product name has multiple Ids attached to it 
the probleme is that we kan give a product name multiple id because it might have the same name a chair but the color is different for example ared chair has id1 and a black chair has id2 which is plausible with a one name to elemnate the confusion we decided to create a new column contains a unique name  that solves the probleme 

In [43]:
SuperStore["Unique Product Name"]=SuperStore["Product Name"]+"("+SuperStore["Product ID"]+")"
SuperStore["Product ID"].nunique()-SuperStore["Unique Product Name"].nunique()

0

at this point our dataset is valid the number of names equales the number of IDs for all the products 

### Negative or zero value 
we do not have any zero or sub zero values in the numerical colomns 


In [44]:
print(SuperStore[SuperStore["Row ID"] <= 0])
print(SuperStore[SuperStore["Sales"] <= 0])
print((SuperStore["Postal Code"] == "0").sum() ) 

Empty DataFrame
Columns: [Row ID, Order ID, Order Date, Ship Date, Ship Mode, Customer ID, Customer Name, Segment, Country, City, State, Postal Code, Region, Product ID, Category, Sub-Category, Product Name, Sales, Unique Product Name]
Index: []
Empty DataFrame
Columns: [Row ID, Order ID, Order Date, Ship Date, Ship Mode, Customer ID, Customer Name, Segment, Country, City, State, Postal Code, Region, Product ID, Category, Sub-Category, Product Name, Sales, Unique Product Name]
Index: []
0


### Date inconsistencies 
we will turn the dates from strings to datetime object so the manipulation is easiere 

In [45]:
SuperStore["Order Date"]=pd.to_datetime(SuperStore["Order Date"] , errors="coerce" , dayfirst=True)
SuperStore["Ship Date"]=pd.to_datetime(SuperStore["Ship Date"] , errors="coerce" , dayfirst=True)

print(SuperStore["Order Date"].isna().sum())
print(SuperStore["Ship Date"].isna().sum())


0
0


Now we check for logical inconsistencies such as 

shippin date must always be > than order Date 

minimum and maximum values

In [46]:
TimeO=SuperStore[SuperStore["Ship Date"] < SuperStore ["Order Date"]]
print((TimeO))
print(SuperStore["Ship Date"].min())
print(SuperStore["Ship Date"].max())
print(SuperStore["Order Date"].min())
print(SuperStore["Order Date"].max())

Empty DataFrame
Columns: [Row ID, Order ID, Order Date, Ship Date, Ship Mode, Customer ID, Customer Name, Segment, Country, City, State, Postal Code, Region, Product ID, Category, Sub-Category, Product Name, Sales, Unique Product Name]
Index: []
2015-01-07 00:00:00
2019-01-05 00:00:00
2015-01-03 00:00:00
2018-12-30 00:00:00


In [47]:
SuperStore.info()

<class 'pandas.DataFrame'>
Index: 9799 entries, 0 to 9799
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Row ID               9799 non-null   int64         
 1   Order ID             9799 non-null   str           
 2   Order Date           9799 non-null   datetime64[us]
 3   Ship Date            9799 non-null   datetime64[us]
 4   Ship Mode            9799 non-null   str           
 5   Customer ID          9799 non-null   str           
 6   Customer Name        9799 non-null   str           
 7   Segment              9799 non-null   str           
 8   Country              9799 non-null   str           
 9   City                 9799 non-null   str           
 10  State                9799 non-null   str           
 11  Postal Code          9799 non-null   str           
 12  Region               9799 non-null   str           
 13  Product ID           9799 non-null   str         